recipes/full_finetune_distributed.py

```diff
@@ -404,15 +404,18 @@ def _setup_model(
         # Shard transformer decoder layers (or AC-wrapped versions)
         # Alternatively we could condition on the module type (TransformerDecoder or CheckpointWrapper)
         # But directly using the name is more concise
-        def _is_layer_fqn(s: str) -> bool:
+        def _is_layer_name(name: str, module: nn.Module) -> bool:
             """
             Return True for layers.i and False for all other module names
             Covers sharding for both AC-wrapped and non-AC-wrapped modules in one shot
             """
-            s_list = s.split(".")
-            return len(s_list) == 2 and s_list[0] == "layers" and str.isdigit(s_list[1])
+            name_list = name.split(".")
```

Comment:
1/ Should we not be using some sort of struct / dataclass to represent layer names consistently across the code so that such bugs can be prevented?

2/ Can we at least move this logic to a util that can be used across recipes rather than having the same logic in 4 places?

Reply:
Yeah agreed on both of these points. On the first one, I think the proper thing to do is to use the module rather than the layer name. Activation checkpointing makes this a bit of a headache cause it modifies the module (so we can't just do something like `isinstance(m, TransformerDecoderLayer)`). So actually this string split version was just a hack to handle both cases in one go and is (clearly as evidenced by this very bug) not the scalable way to do it

Reply:
2/ 100%, i think we have other opportunities like this in our recipes, to reduce tech debt. I will add the utility tomorrow morning.

1/ this is harder. If we have a robust way to tests all of our flags with different models/recipes, that would be easy. But as it is now, i would have to test them manually.